File: src/ugv_teleop/src/keyboard_gui_teleop_node.cpp
#include <algorithm>
#include <atomic>
#include <chrono>
#include <functional>
#include <memory>
#include <mutex>
#include <set>
#include <string>
#include <thread>
#include <vector>

#include <QApplication>
#include <QCloseEvent>
#include <QFocusEvent>
#include <QFontDatabase>
#include <QKeyEvent>
#include <QLabel>
#include <QTimer>
#include <QVBoxLayout>
#include <QWidget>

#include "rcl_interfaces/msg/parameter_descriptor.hpp"
#include "rclcpp/rclcpp.hpp"
#include "ugv_teleop/keyboard_teleop_common.hpp"

namespace ugv_teleop
{

namespace
{
std::string join_keys(const std::vector<std::string> & keys)
{
  if (keys.empty()) {
    return "-";
  }
  std::string joined;
  for (const auto & key : keys) {
    joined += key;
  }
  return joined;
}

std::string normalize_gui_key(const QKeyEvent * event)
{
  if (event == nullptr) {
    return "";
  }

  const QString text = event->text();
  if (!text.isEmpty()) {
    const QChar c = text.at(0).toLower();
    if (c.isLetter()) {
      return std::string(1, static_cast<char>(c.toLatin1()));
    }
  }

  switch (event->key()) {
    case Qt::Key_Q: return "q";
    case Qt::Key_W: return "w";
    case Qt::Key_E: return "e";
    case Qt::Key_A: return "a";
    case Qt::Key_S: return "s";
    case Qt::Key_D: return "d";
    case Qt::Key_X: return "x";
    case Qt::Key_I: return "i";
    case Qt::Key_O: return "o";
    case Qt::Key_K: return "k";
    case Qt::Key_L: return "l";
    default: return "";
  }
}
}

class KeyboardGuiTeleopNode : public rclcpp::Node
{
public:
  KeyboardGuiTeleopNode()
  : rclcpp::Node("keyboard_gui_teleop_node")
  {
    common_params_ = declare_common_parameters(*this);
    core_ = create_core_from_params(common_params_);
    publisher_ = create_twist_publisher_adapter(*this, common_params_);

    rcl_interfaces::msg::ParameterDescriptor read_only_descriptor;
    read_only_descriptor.read_only = true;
    window_title_ = this->declare_parameter<std::string>(
      "window_title", "UGV Keyboard Teleop", read_only_descriptor);
    focus_loss_zero_cmd_ = this->declare_parameter<bool>(
      "focus_loss_zero_cmd", true, read_only_descriptor);

    const auto publish_period = std::chrono::duration<double>(
      1.0 / std::max(1.0, common_params_.publish_rate));
    publish_timer_ = this->create_wall_timer(
      publish_period,
      std::bind(&KeyboardGuiTeleopNode::publish_timer_callback, this));

    status_text_ = k_common_keyboard_message;
  }

  ~KeyboardGuiTeleopNode() override
  {
    shutdown();
  }

  std::string window_title() const
  {
    return window_title_;
  }

  void handle_key_press(const std::string & key)
  {
    if (key.empty()) {
      return;
    }
    std::lock_guard<std::mutex> lock(pressed_keys_mutex_);
    if (pressed_keys_.count(key) > 0U) {
      return;
    }
    if (core_->handle_key_press(key, std::chrono::steady_clock::now())) {
      pressed_keys_.insert(key);
    }
  }

  void handle_key_release(const std::string & key)
  {
    if (key.empty()) {
      return;
    }
    std::lock_guard<std::mutex> lock(pressed_keys_mutex_);
    pressed_keys_.erase(key);
    core_->handle_key_release(key, std::chrono::steady_clock::now());
  }

  void handle_focus_lost()
  {
    {
      std::lock_guard<std::mutex> lock(pressed_keys_mutex_);
      pressed_keys_.clear();
    }
    core_->clear_move_keys(std::chrono::steady_clock::now());
    if (focus_loss_zero_cmd_) {
      publisher_->publish_zero(core_.get());
    }
  }

  std::string status_text() const
  {
    std::lock_guard<std::mutex> lock(status_mutex_);
    return status_text_;
  }

  void shutdown()
  {
    if (shutdown_done_.exchange(true)) {
      return;
    }
    core_->emergency_stop(std::chrono::steady_clock::now());
    if (rclcpp::ok()) {
      publisher_->publish_zero(core_.get());
    }
  }

private:
  void publish_timer_callback()
  {
    const auto command = core_->snapshot(std::chrono::steady_clock::now());
    publisher_->publish(command);
    update_status_text(command);
  }

  void update_status_text(const TeleopCommand & command)
  {
    const std::string text =
      std::string(k_common_keyboard_message) +
      "\nCurrent speed: linear " + std::to_string(command.linear_speed) +
      " m/s | angular " + std::to_string(command.angular_speed) + " rad/s\n" +
      "Current keys: " + join_keys(command.active_keys) + "\n" +
      "Focus loss will stop motion";
    std::lock_guard<std::mutex> lock(status_mutex_);
    status_text_ = text;
  }

  CommonTeleopParams common_params_;
  std::shared_ptr<KeyboardTeleopCore> core_;
  std::shared_ptr<TwistPublisherAdapter> publisher_;
  rclcpp::TimerBase::SharedPtr publish_timer_;

  std::string window_title_;
  bool focus_loss_zero_cmd_{true};

  mutable std::mutex status_mutex_;
  std::string status_text_;

  std::mutex pressed_keys_mutex_;
  std::set<std::string> pressed_keys_;
  std::atomic<bool> shutdown_done_{false};
};

class KeyboardTeleopWindow : public QWidget
{
public:
  explicit KeyboardTeleopWindow(std::shared_ptr<KeyboardGuiTeleopNode> teleop_node)
  : teleop_node_(std::move(teleop_node))
  {
    this->setWindowTitle(QString::fromStdString(teleop_node_->window_title()));
    this->setFixedSize(460, 280);
    this->setFocusPolicy(Qt::StrongFocus);

    auto * layout = new QVBoxLayout(this);
    layout->setContentsMargins(16, 16, 16, 16);

    status_label_ = new QLabel(this);
    status_label_->setAlignment(Qt::AlignTop | Qt::AlignLeft);
    status_label_->setWordWrap(true);
    QFont font = QFontDatabase::systemFont(QFontDatabase::FixedFont);
    font.setPointSize(11);
    status_label_->setFont(font);
    layout->addWidget(status_label_);

    status_timer_ = new QTimer(this);
    connect(status_timer_, &QTimer::timeout, this, &KeyboardTeleopWindow::refresh_status);
    status_timer_->start(100);

    ros_alive_timer_ = new QTimer(this);
    connect(ros_alive_timer_, &QTimer::timeout, this, &KeyboardTeleopWindow::check_ros_alive);
    ros_alive_timer_->start(100);

    refresh_status();
  }

protected:
  void keyPressEvent(QKeyEvent * event) override
  {
    if (event->isAutoRepeat()) {
      event->accept();
      return;
    }
    const std::string key = normalize_gui_key(event);
    if (!key.empty()) {
      teleop_node_->handle_key_press(key);
    }
    QWidget::keyPressEvent(event);
  }

  void keyReleaseEvent(QKeyEvent * event) override
  {
    if (event->isAutoRepeat()) {
      event->accept();
      return;
    }
    const std::string key = normalize_gui_key(event);
    if (!key.empty()) {
      teleop_node_->handle_key_release(key);
    }
    QWidget::keyReleaseEvent(event);
  }

  void focusOutEvent(QFocusEvent * event) override
  {
    teleop_node_->handle_focus_lost();
    QWidget::focusOutEvent(event);
  }

  void closeEvent(QCloseEvent * event) override
  {
    teleop_node_->shutdown();
    if (rclcpp::ok()) {
      rclcpp::shutdown();
    }
    QWidget::closeEvent(event);
  }

private:
  void refresh_status()
  {
    status_label_->setText(QString::fromStdString(teleop_node_->status_text()));
  }

  void check_ros_alive()
  {
    if (!rclcpp::ok()) {
      this->close();
    }
  }

  std::shared_ptr<KeyboardGuiTeleopNode> teleop_node_;
  QLabel * status_label_{nullptr};
  QTimer * status_timer_{nullptr};
  QTimer * ros_alive_timer_{nullptr};
};

}

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  int exit_code = 0;

  std::shared_ptr<ugv_teleop::KeyboardGuiTeleopNode> teleop_node;
  std::thread spinner;

  try {
    teleop_node = std::make_shared<ugv_teleop::KeyboardGuiTeleopNode>();
    spinner = std::thread(
      [teleop_node]() {
        rclcpp::spin(teleop_node);
      });

    QApplication app(argc, argv);
    ugv_teleop::KeyboardTeleopWindow window(teleop_node);
    window.show();
    window.activateWindow();
    window.raise();
    window.setFocus();

    exit_code = app.exec();
  } catch (const std::exception & exception) {
    RCLCPP_ERROR(
      rclcpp::get_logger("keyboard_gui_teleop_node"),
      "Unhandled exception: %s",
      exception.what());
    exit_code = 1;
  }

  if (teleop_node) {
    teleop_node->shutdown();
  }
  if (rclcpp::ok()) {
    rclcpp::shutdown();
  }
  if (spinner.joinable()) {
    spinner.join();
  }
  return exit_code;
}

File: src/ugv_teleop/ugv_teleop/keyboard_gui_teleop_node.py
#!/usr/bin/env python3
import threading
import time

import rcl_interfaces.msg
import rclpy

from .keyboard_teleop_common import (
    COMMON_KEYBOARD_MESSAGE,
    create_core_from_params,
    create_twist_publisher_adapter,
    declare_common_parameters,
)


class KeyboardGuiTeleopNode:
    def __init__(self):
        self.node = rclpy.create_node("keyboard_gui_teleop_node")
        self.common_params = declare_common_parameters(self.node)
        read_only_descriptor = rcl_interfaces.msg.ParameterDescriptor(read_only=True)
        self.window_title = str(
            self.node.declare_parameter(
                "window_title",
                "UGV Keyboard Teleop",
                read_only_descriptor,
            ).value
        )
        self.focus_loss_zero_cmd = bool(
            self.node.declare_parameter(
                "focus_loss_zero_cmd",
                True,
                read_only_descriptor,
            ).value
        )
        self.core = create_core_from_params(self.common_params)
        self.publisher = create_twist_publisher_adapter(self.node, self.common_params)
        self._status_lock = threading.Lock()
        self._status_text = ""
        self._pressed_keys: set[str] = set()

        publish_period = 1.0 / max(1.0, float(self.common_params["publish_rate"]))
        self.node.create_timer(publish_period, self._publish_timer_callback)

    def _publish_timer_callback(self) -> None:
        command = self.core.snapshot(time.monotonic())
        self.publisher.publish(command)
        self._update_status_text(command)

    def _update_status_text(self, command) -> None:
        active_keys = "".join(command.active_keys) if command.active_keys else "-"
        text = (
            f"{COMMON_KEYBOARD_MESSAGE.strip()}\n\n"
            f"当前速度: 线速度 {command.linear_speed:.2f} m/s | 角速度 {command.angular_speed:.2f} rad/s\n"
            f"当前按键: {active_keys}\n"
            f"窗口失焦会立即停车"
        )
        with self._status_lock:
            self._status_text = text

    def get_status_text(self) -> str:
        with self._status_lock:
            return self._status_text

    def handle_key_press(self, key: str) -> None:
        if not key:
            return
        normalized = key.lower()
        if normalized in self._pressed_keys:
            return
        if self.core.handle_key_press(normalized, time.monotonic()):
            self._pressed_keys.add(normalized)

    def handle_key_release(self, key: str) -> None:
        if not key:
            return
        normalized = key.lower()
        self._pressed_keys.discard(normalized)
        self.core.handle_key_release(normalized, time.monotonic())

    def handle_focus_lost(self) -> None:
        self._pressed_keys.clear()
        self.core.clear_move_keys(time.monotonic())
        if self.focus_loss_zero_cmd:
            self.publisher.publish_zero(self.core)

    def shutdown(self) -> None:
        self.core.emergency_stop(time.monotonic())
        self.publisher.publish_zero(self.core)
        self.node.destroy_node()


def main():
    try:
        import tkinter as tk
    except ImportError:
        print("tkinter 不可用，请改用 keyboard_backend:=tty")
        return

    rclpy.init()
    teleop = KeyboardGuiTeleopNode()
    spinner = threading.Thread(target=rclpy.spin, args=(teleop.node,), daemon=True)
    spinner.start()

    try:
        root = tk.Tk()
    except tk.TclError as exc:
        teleop.node.get_logger().error(f"无法启动键盘 GUI teleop: {exc}")
        teleop.shutdown()
        rclpy.shutdown()
        spinner.join(timeout=1.0)
        return

    status_var = tk.StringVar(value=COMMON_KEYBOARD_MESSAGE.strip())

    def on_key_press(event):
        teleop.handle_key_press(event.char or event.keysym)

    def on_key_release(event):
        teleop.handle_key_release(event.char or event.keysym)

    def on_focus_out(_event):
        teleop.handle_focus_lost()

    def refresh_status():
        status_var.set(teleop.get_status_text())
        if root.winfo_exists():
            root.after(100, refresh_status)

    def close_window():
        teleop.shutdown()
        if rclpy.ok():
            rclpy.shutdown()
        root.destroy()

    root.title(teleop.window_title)
    root.geometry("460x280")
    root.resizable(False, False)
    root.bind("<KeyPress>", on_key_press)
    root.bind("<KeyRelease>", on_key_release)
    root.bind("<FocusOut>", on_focus_out)
    root.protocol("WM_DELETE_WINDOW", close_window)
    root.configure(padx=16, pady=16)

    label = tk.Label(
        root,
        textvariable=status_var,
        justify="left",
        anchor="nw",
        font=("TkFixedFont", 11),
    )
    label.pack(fill="both", expand=True)

    root.focus_force()
    refresh_status()
    try:
        root.mainloop()
    finally:
        if rclpy.ok():
            teleop.shutdown()
            rclpy.shutdown()
        spinner.join(timeout=1.0)


if __name__ == "__main__":
    main()
